src/coreclr/vm/threadstatics.cpp

```diff
@@ -51,14 +51,17 @@ void ThreadLocalBlock::FreeTLM(SIZE_T i, BOOL isThreadShuttingdown)
                     {
                         ThreadLocalModule::CollectibleDynamicEntry *entry = (ThreadLocalModule::CollectibleDynamicEntry*)pThreadLocalModule->m_pDynamicClassTable[k].m_pDynamicEntry;
                         PTR_LoaderAllocator pLoaderAllocator = entry->m_pLoaderAllocator;
-
-                        if (entry->m_hGCStatics != NULL)
-                        {
-                            pLoaderAllocator->FreeHandle(entry->m_hGCStatics);
-                        }
-                        if (entry->m_hNonGCStatics != NULL)
+                        // Skip assemblies that are already unloaded
+                        if (!pLoaderAllocator->IsUnloaded())
```

Comment:
Can we run into a situation where the thread gets rescheduled right after this check, the code in GCLoaderAllocators sets the flag and start destroying the loader allocators, and then the thread running this code gets scheduled back and still hits this crash?

Reply:
good question! I think it may happen and ideally `FreeHandle` call of LoaderAllocator itself should deal with its own state then, perhaps we could use `m_crstLoaderAllocator` lock, but it would mean some perf hit for the fast pointer path

my ideal expectation would be that since `DeleteThreadStaticData` is under `GCX_COOP` in `Thread::OnThreadTerminate` that should hopefully mean that once we are in `DeleteThreadStaticData` call we know for sure whether or not `LoaderAllocator` is alive or not and that state can not change - perhaps `IsAlive()` should be used as `IsUnloaded` is set on finalizer thread and not relevant to `GCX_COOP` 🤔